packages/node-core/src/indexer/project.service.spec.ts

```diff
@@ -0,0 +1,167 @@
+    expect(result).toBe(true);
+  });
+
+  it('hasDataSourcesAfterHeight - undefined endBlock', async () => {
+    service.getDataSources = jest.fn().mockResolvedValue([{endBlock: 100}, {endBlock: undefined}]);
+
+    const result = await service.hasDataSourcesAfterHeight(150);
+    expect(result).toBe(true);
+  });
+
+  it('getDataSources', async () => {
+    (service as any).project.dataSources = [
+      {startBlock: 100, endBlock: 200},
+      {startBlock: 1, endBlock: 100},
+    ];
+    (service as any).dynamicDsService.getDynamicDatasources = jest
+      .fn()
+      .mockResolvedValue([{startBlock: 150, endBlock: 250}]);
+
+    const result = await service.getDataSources(175);
+    expect(result).toEqual([
+      {startBlock: 100, endBlock: 200},
+      {startBlock: 150, endBlock: 250},
+    ]);
+  });
+
+  describe('getDatasourceMap', () => {
+    it('should add endBlock heights correctly', () => {
+      (service as any).dynamicDsService.dynamicDatasources = [];
+      (service as any).projectUpgradeService = {
+        projects: [
+          [
+            1,
+            {
+              dataSources: [
+                {startBlock: 1, endBlock: 300},
+                {startBlock: 10, endBlock: 20},
+                {startBlock: 1, endBlock: 100},
+                {startBlock: 50, endBlock: 200},
+                {startBlock: 500},
+              ],
+            },
+          ],
+        ],
+      } as any;
+
+      const result = service.getDataSourcesMap();
+      expect(result.getAll()).toEqual(
+        new Map([
+          [
+            1,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+            ],
+          ],
+          [
+            10,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+              {startBlock: 10, endBlock: 20},
+            ],
+          ],
+          [
+            21,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+            ],
+          ],
+          [
+            50,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 1, endBlock: 100},
+              {startBlock: 50, endBlock: 200},
+            ],
+          ],
+          [
+            101,
+            [
+              {startBlock: 1, endBlock: 300},
+              {startBlock: 50, endBlock: 200},
+            ],
+          ],
+          [201, [{startBlock: 1, endBlock: 300}]],
+          [301, []],
+          [500, [{startBlock: 500}]],
+        ])
+      );
+    });
+
+    it('should contain datasources from current project only', () => {
+      (service as any).dynamicDsService.dynamicDatasources = [];
+      (service as any).projectUpgradeService = {
+        projects: [
+          [
+            1,
+            {
+              dataSources: [{startBlock: 1}],
```

Comment:
This is a pre-existing bug buy if this change is made then it will produce a different result but the expected result is as it is now
```suggestion
              dataSources: [{startBlock: 1}, {startBlock: 200}],
```